### PR TITLE
Checks ref count and slot list again before flushing index entry

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

When flushing the in-mem accounts index, there's a race from when we check if the entry should be evicted to when we actually do the eviction. If the foreground *did* update the entry such that it no longer had ref count == 1 && slot list == 1, we would still evict it. This would cause the disk index to see a multi value (or none) entry, and then create new data files.

#### Summary of Changes

Check the ref count and slot list again before flushing the index entry.

### CODE REVIEW

**Review by jeffwashington — APPROVED** [review]
this is perfect. ty!

File: accounts-db/src/accounts_index/in_mem_accounts_index.rs [hunk 1]
```diff
@@ -1255,14 +1255,24 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> InMemAccountsIndex<T,
...
                                         .take()
                                         .unwrap_or_else(|| v.slot_list.read().unwrap());
+                                    // Check the ref count and slot list one more time before flushing.
+                                    // It is possible the foreground has updated this entry since
+                                    // we last checked above in `should_evict_from_mem()`.
+                                    // If the entry *was* updated, re-mark it as dirty then
+                                    // skip to the next pubkey/entry.
+                                    let ref_count = v.ref_count();
+                                    if ref_count != 1 || slot_list.len() != 1 {
+                                        v.set_dirty(true);
```
Comment (HaoranYi) [inline accounts-db/src/accounts_index/in_mem_accounts_index.rs:1265]: Will it be useful to add a stat here to see how often this actually happens?

**Review by HaoranYi — APPROVED** [review]
lgtm!

File: accounts-db/src/accounts_index/in_mem_accounts_index.rs (hunk 1 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_index/in_mem_accounts_index.rs:1265]: i would be curious how often it happens. Could give us indications of how often we may have been seeing this race. I think a metric is good. we dump index things every 10s.

Comment (jeffwashington) [reply accounts-db/src/accounts_index/in_mem_accounts_index.rs:1265]: can be added after. This is important to put in. I don't think this is urgent enough to backport. It isn't that common and the system still works fine.

Comment (jeffwashington) [reply accounts-db/src/accounts_index/in_mem_accounts_index.rs:1265]: once clean or shrink runs, the entry will be modified, eventually settle on happy values and get written back to disk.

**Pull request merged**